# Regression Model Validation

## Introduction

Previously you've evaluated a multiple linear regression model by calculating metrics based on the fit of the training data. In this lesson you'll learn why it's important to split your data in a train and a test set if you want to evaluate a model used for prediction.

## Objectives

You will be able to:

* Perform a train-test split
* Prepare training and testing data for modeling
* Compare training and testing errors to determine if model is over or underfitting

## Model Evaluation

Recall some ways that we can evaluate linear regression models.

### Residuals

It is pretty straightforward that, to evaluate the model, you'll want to compare your predicted values, $\hat y$ with the actual value, $y$. The difference between the two values is referred to as the **residuals**:

$r_{i} = y_{i} - \hat y_{i}$ 

To get a summarized measure over all the instances, a popular metric is the (Root) Mean Squared Error:

RMSE = $\sqrt{\frac{1}{n}\sum_{i=1}^{n}(y_{i} - \hat y_{i})^2}$

MSE = $\frac{1}{n}\sum_{i=1}^{n}(y_{i} - \hat y_{i})^2$

Larger (R)MSE indicates a _worse_ model fit.

## The Need for Train-Test Split

### Making Predictions and Evaluation

So far we've simply been fitting models to data, and evaluated our models calculating the errors between our $\hat y$ and our actual targets $y$, while these targets $y$ contributed in fitting the model.

Let's say we want to predict the outcome for observations that are not necessarily in our dataset now; e.g: we want to **predict** miles per gallon for a new car that isn't part of our dataset, or predict the price for a new house in Ames.

In order to get a good sense of how well your model will be doing on new instances, you'll have to perform a so-called "train-test-split". What you'll be doing here, is taking a sample of the data that serves as input to "train" our model - fit a linear regression and compute the parameter estimates for our variables, and calculate how well our predictive performance is doing comparing the actual targets $y$ and the fitted $\hat y$ obtained by our model.

### Underfitting and Overfitting

Another reason to use train-test-split is because of a common problem which doesn't only affect linear models, but nearly all (other) machine learning algorithms: overfitting and underfitting. An overfit model is not generalizable and will not hold to future cases. An underfit model does not make full use of the information available and produces weaker predictions than is feasible. The following image gives a nice, more general demonstration:

<img src='https://curriculum-content.s3.amazonaws.com/data-science/images/new_overfit_underfit.png'>

## Mechanics of Train-Test Split

When performing a train-test-split, it is important that the data is **randomly** split. At some point, you will encounter datasets that have certain characteristics that are only present in certain segments of the data. For example, if you were looking at sales data for a website, you might expect the data to look different on days that promotional deals were held versus days that deals were not held. If we don't randomly split the data, there is a chance we might overfit to the characteristics of certain segments of data.

Another thing to consider is just **how big** each training and testing set should be. There is no hard and fast rule for deciding the correct size, but the range of training set is usually anywhere from 66% - 80% (and testing set between 33% and 20%). Some types of machine learning models need a substantial amount of data to train on, and as such, the training sets should be larger. Some models with many different tuning parameters will need to be validated with larger sets (the test size should be larger) to determine what the optimal parameters should be. When in doubt, just stick with training set sizes around 70% and test set sizes around 30%.

## Train-Test Split with Scikit-Learn

You could write your own pandas code to shuffle and split your data, but we'll use the convenient `train_test_split` function from scikit-learn instead. We'll also use the Auto MPG dataset.

In [1]:
import pandas as pd

data = pd.read_csv('auto-mpg.csv')
data

mpg  cylinders  displacement  horsepower  weight  acceleration   
0    18.0          8         307.0         130    3504          12.0  \
1    15.0          8         350.0         165    3693          11.5   
2    18.0          8         318.0         150    3436          11.0   
3    16.0          8         304.0         150    3433          12.0   
4    17.0          8         302.0         140    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
387  27.0          4         140.0          86    2790          15.6   
388  44.0          4          97.0          52    2130          24.6   
389  32.0          4         135.0          84    2295          11.6   
390  28.0          4         120.0          79    2625          18.6   
391  31.0          4         119.0          82    2720          19.4   

     model year  origin                   car name  
0            70       1  chevrolet chevelle malibu  
1            70       1          buick skylark 320  
2            70       1         plymouth satellite  
3            70       1              amc rebel sst  
4            70       1                ford torino  
..          ...     ...                        ...  
387          82       1            ford mustang gl  
388          82       2                  vw pickup  
389          82       1              dodge rampage  
390          82       1                ford ranger  
391          82       1                 chevy s-10  

[392 rows x 9 columns]

The `train_test_split` function ([documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)) takes in a series of array-like variables, as well as some optional arguments. It returns multiple arrays.

For example, this would be a valid way to use `train_test_split`:

In [2]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(data)

In [3]:
train

mpg  cylinders  displacement  horsepower  weight  acceleration   
217  25.5          4         122.0          96    2300          15.5  \
211  13.0          8         350.0         145    4055          12.0   
60   21.0          4         122.0          86    2226          16.5   
203  28.0          4          97.0          75    2155          16.4   
198  18.0          6         250.0          78    3574          21.0   
..    ...        ...           ...         ...     ...           ...   
298  23.9          8         260.0          90    3420          22.2   
175  23.0          4         115.0          95    2694          15.0   
355  30.7          6         145.0          76    3160          19.6   
93   13.0          8         440.0         215    4735          11.0   
271  23.9          4         119.0          97    2405          14.9   

     model year  origin                           car name  
217          77       1                  plymouth arrow gs  
211          76       1                          chevy c10  
60           72       1                ford pinto runabout  
203          76       3                     toyota corolla  
198          76       1                  ford granada ghia  
..          ...     ...                                ...  
298          79       1  oldsmobile cutlass salon brougham  
175          75       2                         audi 100ls  
355          81       2                       volvo diesel  
93           73       1       chrysler new yorker brougham  
271          78       3                      datsun 200-sx  

[294 rows x 9 columns]

In [4]:
test

mpg  cylinders  displacement  horsepower  weight  acceleration   
135  13.0          8         350.0         150    4699          14.5  \
209  16.5          6         168.0         120    3820          16.7   
275  16.2          6         163.0         133    3410          15.8   
125  19.0          6         232.0         100    2901          16.0   
71   15.0          8         304.0         150    3892          12.5   
..    ...        ...           ...         ...     ...           ...   
322  40.8          4          85.0          65    2110          19.2   
82   28.0          4          98.0          80    2164          15.0   
65   17.0          8         304.0         150    3672          11.5   
137  14.0          8         302.0         140    4638          16.0   
385  36.0          4         135.0          84    2370          13.0   

     model year  origin                  car name  
135          74       1  buick century luxus (sw)  
209          76       2        mercedes-benz 280s  
275          78       2             peugeot 604sl  
125          74       1                amc hornet  
71           72       1          amc matador (sw)  
..          ...     ...                       ...  
322          80       3                datsun 210  
82           72       1           dodge colt (sw)  
65           72       1        amc ambassador sst  
137          74       1     ford gran torino (sw)  
385          82       1         dodge charger 2.2  

[98 rows x 9 columns]

In this case, the DataFrame `data` was split into two DataFrames called `train` and `test`. `train` has 294 values (75% of the full dataset) and `test` has 98 values (25% of the full dataset). Note the randomized order of the index values on the left.

However you can also pass multiple array-like variables into `train_test_split` at once. For each variable that you pass in, you will get a train and a test copy back out.

Most commonly in this curriculum these are the inputs and outputs:

Inputs

- `X`
- `y`

Outputs

- `X_train`
- `X_test`
- `y_train`
- `y_test`

In [5]:
y = data[['mpg']]
X = data.drop(['mpg', 'car name'], axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y)

In [6]:
X_train

cylinders  displacement  horsepower  weight  acceleration  model year   
294          4         121.0          80    2670          15.0          79  \
240          4         121.0         110    2600          12.8          77   
242          4          90.0          48    1985          21.5          78   
39           8         351.0         153    4154          13.5          71   
125          6         232.0         100    2901          16.0          74   
..         ...           ...         ...     ...           ...         ...   
59           4         140.0          90    2408          19.5          72   
137          8         302.0         140    4638          16.0          74   
369          4         105.0          74    1980          15.3          82   
174          6         232.0          90    3211          17.0          75   
198          6         250.0          78    3574          21.0          76   

     origin  
294       1  
240       2  
242       2  
39        1  
125       1  
..      ...  
59        1  
137       1  
369       2  
174       1  
198       1  

[294 rows x 7 columns]

In [7]:
y_train

mpg
294  27.4
240  21.5
242  43.1
39   14.0
125  19.0
..    ...
59   20.0
137  14.0
369  36.0
174  19.0
198  18.0

[294 rows x 1 columns]

We can view the lengths of the results like this:

In [8]:
print(len(X_train), len(X_test), len(y_train), len(y_test))

294 98 294 98


However it is not recommended to pass in just the data to be split. This is because the randomization of the split means that you will get different results for `X_train` etc. every time you run the code. **For reproducibility, it is always recommended that you specify a `random_state`**, such as in this example:

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

Another optional argument is `test_size`, which makes it possible to choose the size of the test set and the training set instead of using the default 75% train/25% test proportions.

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Note that the lengths of the resulting datasets will be different:

In [11]:
print(len(X_train), len(X_test), len(y_train), len(y_test))

313 79 313 79


## Preparing Data for Modeling

When using a train-test split, data preparation should happen _after_ the split. This is to avoid ***data leakage***. The general idea is that the treatment of the test data should be as similar as possible to how genuinely unknown data should be treated. And genuinely unknown data would not have been there at the time of fitting the scikit-learn transformers, just like it would not have been there at the time of fitting the model!

In some cases you will see all of the data being prepared together for expediency, but the best practice is to prepare it separately.

### Log Transformation

In [12]:
from sklearn.preprocessing import FunctionTransformer
import numpy as np

# Instantiate a custom transformer for log transformation 
log_transformer = FunctionTransformer(np.log, validate=True)

# Columns to be log transformed 
log_columns = ['displacement', 'horsepower', 'weight']

# New names for columns after transformation
new_log_columns = ['log_disp', 'log_hp', 'log_wt']

# Log transform the training columns and convert them into a DataFrame 
X_train_log = pd.DataFrame(log_transformer.fit_transform(X_train[log_columns]), 
                           columns=new_log_columns, index=X_train.index)

# Replace training columns with transformed versions
X_train = pd.concat([X_train.drop(log_columns, axis=1), X_train_log], axis=1)
X_train

cylinders  acceleration  model year  origin  log_disp    log_hp    log_wt
258          6          18.7          78       1  5.416100  4.700480  8.194229
182          4          14.9          76       1  4.941642  4.521789  7.852439
172          6          14.5          75       1  5.141664  4.574711  8.001020
63           8          13.5          72       1  5.762051  5.010635  8.327243
340          4          16.4          81       1  4.454347  4.158883  7.536364
..         ...           ...         ...     ...       ...       ...       ...
71           8          12.5          72       1  5.717028  5.010635  8.266678
106          6          15.0          73       1  5.446737  4.605170  7.933438
270          4          17.6          78       1  5.017280  4.442651  7.956827
348          4          20.7          81       1  4.584967  4.174387  7.774856
102          8          14.0          73       1  5.991465  5.010635  8.516593

[313 rows x 7 columns]

In [13]:
# Log transform the test columns and convert them into a DataFrame 
X_test_log = pd.DataFrame(log_transformer.transform(X_test[log_columns]), 
                          columns=new_log_columns, index=X_test.index)

# Replace testing columns with transformed versions
X_test = pd.concat([X_test.drop(log_columns, axis=1), X_test_log], axis=1)
X_test

cylinders  acceleration  model year  origin  log_disp    log_hp    log_wt
78           4          18.0          72       2  4.564348  4.234107  7.691200
274          4          15.7          78       2  4.795791  4.744932  7.935587
246          4          16.4          78       3  4.510860  4.094345  7.495542
55           4          20.5          71       1  4.510860  4.248495  7.578145
387          4          15.6          82       1  4.941642  4.454347  7.933797
..         ...           ...         ...     ...       ...       ...       ...
361          6          16.6          81       1  5.416100  4.442651  8.150468
82           4          15.0          72       1  4.584967  4.382027  7.679714
114          8          13.0          73       1  5.857933  4.976734  8.314342
3            8          12.0          70       1  5.717028  5.010635  8.141190
18           4          14.5          70       3  4.574711  4.477337  7.663877

[79 rows x 7 columns]

### One-Hot Encoding

In [14]:
from sklearn.preprocessing import OneHotEncoder

# Instantiate OneHotEncoder
ohe = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

# Create X_cat which contains only the categorical variables
cat_columns = [col for col in X.columns if X[col].dtype in [object]]
X_train_cat = X_train.loc[:, cat_columns]

# Fill missing values with the string 'missing'
X_train_cat.fillna(value='missing', inplace=True)

# Transform training set
X_train_ohe = pd.DataFrame(ohe.fit_transform(X_train_cat),
                           columns=cat_columns, index=X_train.index)

In [15]:
# Transform testing set
X_test_cat = pd.DataFrame(ohe.transform(X_test[cat_columns]),
                           columns=cat_columns, index=X_test.index)

# Fill missing values with the string 'missing'
X_test_cat.fillna(value='missing', inplace=True)

# Transform training set
X_test_ohe = pd.DataFrame(ohe.fit_transform(X_test_cat),
                           columns=cat_columns, index=X_test.index)

## Building, Evaluating, and Validating a Model

Great, now that you have preprocessed all the columns, you can fit a linear regression model: 

In [16]:
from sklearn.linear_model import LinearRegression
linreg = LinearRegression()
linreg.fit(X_train, y_train)

y_hat_train = linreg.predict(X_train)
y_hat_test = linreg.predict(X_test)

Look at the residuals and calculate the MSE for training and test sets:  

In [17]:
train_residuals = y_hat_train - y_train
test_residuals = y_hat_test - y_test

In [18]:
mse_train = np.sum((y_train - y_hat_train)**2)/len(y_train)
mse_test = np.sum((y_test - y_hat_test)**2)/len(y_test)
print('Train Mean Squared Error:', mse_train)
print('Test Mean Squared Error:', mse_test)

Train Mean Squared Error: mpg    9.143784
dtype: float64
Test Mean Squared Error: mpg    10.010717
dtype: float64


You can also do this directly using sklearn's `mean_squared_error()` function: 

In [19]:
from sklearn.metrics import mean_squared_error

train_mse = mean_squared_error(y_train, y_hat_train)
test_mse = mean_squared_error(y_test, y_hat_test)
print('Train Mean Squared Error:', train_mse)
print('Test Mean Squared Error:', test_mse)

Train Mean Squared Error: 9.143783702580818
Test Mean Squared Error: 10.010717292450195


Great, there does not seem to be a big difference between the train and test MSE! Interestingly, the test set error is smaller than the training set error. This is fairly rare but does occasionally happen.

In other words, our validation process has indicated that we are **not** overfitting. In fact, we may be _underfitting_ because linear regression is not a very complex model.

## Overfitting with a Different Model

Just for the sake of example, here is a model that is overfit to the data. Don't worry about the model algorithm being shown! Instead, just look at the MSE for the train vs. test set, using the same preprocessed data:

In [20]:
from sklearn.tree import DecisionTreeRegressor

other_model = DecisionTreeRegressor(random_state=42)
other_model.fit(X_train, y_train)

other_train_mse = mean_squared_error(y_train, other_model.predict(X_train))
other_test_mse = mean_squared_error(y_test, other_model.predict(X_test))
print('Train Mean Squared Error:', other_train_mse)
print('Test Mean Squared Error:', other_test_mse)

Train Mean Squared Error: 0.0
Test Mean Squared Error: 10.84253164556962


This model initially seems great...0 MSE for the training data! But then you see that it is performing much worse than our linear regression model on the test data. This model **is** overfitting.

## Additional Resources

[This blog post](https://towardsdatascience.com/linear-regression-in-python-9a1f5f000606) shows a walkthrough of the key steps for model validation with train-test split and scikit-learn.

## Summary 

In this lesson, you learned the importance of the train-test split approach and used one of the most popular metrics for evaluating regression models, (R)MSE. You also saw how to use the `train_test_split` function from `sklearn` to split your data into training and test sets, and then evaluated whether models were overfitting using metrics on those training and test sets.